### DISCUSSION TITLE
How to add new fields without causing deserialization errors?

### DISCUSSION THREAD

**Discussion by jbksr** [miDeb/digitales_register#19] [General]
I've been working on changing the exam/test parsing in the app. It works perfefctly to use the `homework` value from the API. At least for my school's register this works just fine and even seems to be implemented similarly in the official front-end.

**I've made the following changes:**

1. Modify `Homework` and `HomeworkExam` classes
```
abstract class Homework implements Built<Homework, HomeworkBuilder> {
  ...
  bool get warning;
  ...
}

abstract class HomeworkExam
    implements Built<HomeworkExam, HomeworkExamBuilder> {
  int get id;
  String get name;
  bool get online;
  UtcDateTime get deadline;
  bool get hasGrades;
  bool get hasGradeGroupSubmissions;
  int get typeId;
  String get typeName;
  bool get warning;
  static Serializer<HomeworkExam> get serializer => _$homeworkExamSerializer;
  factory HomeworkExam([Function(HomeworkExamBuilder b)? updates]) =
      _$HomeworkExam;
  HomeworkExam._();
}
```
2. Modify the `_parseHomework` and `_parseHomeworkExam` parse functions:
```
Homework _parseHomework(Map data) {
  return Homework((b) {
    b
      ..id = getInt(data["id"])
      ..title = getString(data["title"])
      ..subtitle = getString(data["subtitle"])
      ..label = getString(data["label"])
      ..warningServerSaid = getBool(data["warning"]) ?? b.warningServerSaid

      // if "homework" is zero, it's an entry with a warning
      ..warning = data["homework"] == 0

      ...
}

HomeworkExam _parseHomeworkExam(Map homeworkExam) {
  return HomeworkExam((b) => b
    ..deadline = UtcDateTime.parse(getString(homeworkExam["deadline"])!)
    ..hasGradeGroupSubmissions =
        getBool(homeworkExam["hasGradeGroupSubmissions"])
    ..hasGrades = getBool(homeworkExam["hasGrades"])

    // if "homework" is zero, it's an entry with a warning
    ..warning = homeworkExam["homework"] == 0

    ...
}
```

3. Changes to adapt the UI in calender_card.dart
```
...
 if (homeworkExam.warning)
    _ContentItem(
         title: homeworkExam.typeName,
         content: homeworkExam.name,
         icon: Icons.grade,
          iconColor: Colors.red,
    )
...
```

4. Remove estimateShouldWarn function

(There probably are some other things left to adapt)

### The problem:
For users coming from an older version where the field didn't exist, it will throw an error while deserializing.

### The solution:
I think the solution is to make the field nullable. But I'm not exactly sure how to achieve that.
Another solution is to clear all stored data beforehand, but that's not the right solution obviously.

Here's the debug log:
```
[log] login succeeded
[log] loading state
[log] Failed to load data
[log] DeserializationError (Deserializing '[AppState, dashboardState, [future, true, blacklist, [], allDays, [[homework,...' to 'unspecified' failed due to: Deserializing '[future, true, blacklist, [], allDays, [[homework, [[deleted, false, id, 2, i...' to 'DashboardState' failed due to: Deserializing '[[homework, [[deleted, false, id, 2, isNew, false, isChanged, false, title, E...' to 'BuiltList<Day>' failed due to: Deserializing '[homework, [[deleted, false, id, 2, isNew, false, isChanged, false, title, Er...' to 'Day' failed due to: Deserializing '[[deleted, false, id, 2, isNew, false, isChanged, false, title, Erinnerung, s...' to 'BuiltList<Homework>' failed due to: Deserializing '[deleted, false, id, 2, isNew, false, isChanged, false, title, Erinnerung, su...' to 'Homework' failed due to: Tried to construct class "Homework" with null for non-nullable field "warning".)
```

**Comment by miDeb** [comment]
Hi, sorry for the slow response. You can make things nullable in Dart by appending a `?` to the type, such as `bool?`. Your suggestion sounds good, I'd be happy to accept a pull request!

**Comment by jbksr** [comment]
Ok, it seems to work quite well. The only issue like I mentioned is, that it will fail to deserialise existing data. If the device is online then that's no issue, it can just overwrite it with the new data.
Considering that the device is most likely online when opening the app first after an update, it's ok I think.
Alternatively, what I've also tested is to make the field nullable. This causes the app to at least show all data, but no tests are marked anymore obviously. This is the smarter approach but then having to write `if (homeworkExam.warning == true)` is just so messy.
Obviously the field would need to fall back to false and not mess up the whole deserialisation process. But how?

I haven't got much experience with GitHub and git and it seems to have added my commit to the previous PR #20. That wasn't my intention, I hope that's fine.

**Comment by jbksr** [comment]
Awesome, I'd say this is solved! Again, thank you so much for the kind support and help! I really appreciate it :)

**Comment by miDeb** [comment]
> I haven't got much experience with GitHub and git and it seems to have added my commit to the previous PR https://github.com/miDeb/digitales_register/pull/20. That wasn't my intention, I hope that's fine.

Just to add a quick explainer, when you open a PR in github you always open it from a branch in your repository. In your case you opened it from the main branch of your fork, which means when I merged it the changes in your main branch were pulled in to the main branch of this repo. When you added another commit to your main branch, it automatically became part of the PR.

Usually you don't open a PR from the main branch, but create a new branch for each PR that has the commits you want to be part of this PR. If you're using VS Code or Android Studio, there should be a button somewhere that allows you create new branches and switch between them. (e.g. in VS Code at the bottom left you see the current branch and by clicking it you can switch branches and create new ones)

Git and Github can be a bit overwhelming at first (guess how I know), but you're only going to improve over time ;)
